kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskFactory.java

```diff
@@ -0,0 +1,260 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.slack.kaldb.metadata.recovery.RecoveryTaskMetadata;
+import com.slack.kaldb.metadata.recovery.RecoveryTaskMetadataStore;
+import com.slack.kaldb.metadata.snapshot.SnapshotMetadata;
+import com.slack.kaldb.metadata.snapshot.SnapshotMetadataStore;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for the indexer startup operations like stale live snapshot cleanup.
+ * determining the start indexing offset from metadata and optionally creating a recovery task etc.
+ */
+public class RecoveryTaskFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(RecoveryTaskFactory.class);
+  private static final int SNAPSHOT_OPERATION_TIMEOUT_SECS = 10;
+
+  private final SnapshotMetadataStore snapshotMetadataStore;
+  private final RecoveryTaskMetadataStore recoveryTaskMetadataStore;
+  private final String partitionId;
+  private final long maxOffsetDelay;
+
+  public static final String STALE_SNAPSHOT_DELETE_SUCCESS = "stale_snapshot_delete_success";
+  public static final String STALE_SNAPSHOT_DELETE_FAILED = "stale_snapshot_delete_failed";
+
+  private final Counter snapshotDeleteSuccess;
+  private final Counter snapshotDeleteFailed;
+
+  public RecoveryTaskFactory(
+      SnapshotMetadataStore snapshotMetadataStore,
+      RecoveryTaskMetadataStore recoveryTaskMetadataStore,
+      String partitionId,
+      long maxOffsetDelay,
+      MeterRegistry meterRegistry) {
+    checkState(
+        partitionId != null && !partitionId.isEmpty(), "partitionId shouldn't be null or empty");
+    checkState(maxOffsetDelay > 0, "maxOffsetDelay should be a positive number");
+    this.snapshotMetadataStore = snapshotMetadataStore;
+    this.recoveryTaskMetadataStore = recoveryTaskMetadataStore;
+    this.partitionId = partitionId;
+    this.maxOffsetDelay = maxOffsetDelay;
+
+    snapshotDeleteSuccess = meterRegistry.counter(STALE_SNAPSHOT_DELETE_SUCCESS);
+    snapshotDeleteFailed = meterRegistry.counter(STALE_SNAPSHOT_DELETE_FAILED);
+  }
+
+  @VisibleForTesting
+  public List<SnapshotMetadata> getStaleLiveSnapshots(List<SnapshotMetadata> snapshots) {
+    return snapshots
+        .stream()
+        .filter(snapshotMetadata -> snapshotMetadata.partitionId.equals(partitionId))
+        .filter(SnapshotMetadata::isLive)
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  @VisibleForTesting
+  public List<SnapshotMetadata> deleteStaleLiveSnapshots(List<SnapshotMetadata> snapshots) {
```

Comment:
Why not just reuse `deleteSnapahots` and the caller can call `getStaleLiveSnapshots` to get the list of stale snapshots if that's what it wants to do?